source/src/ble/cs_CrownstoneCentral.cpp

```diff
@@ -0,0 +1,663 @@
+/*
+ * Author: Crownstone Team
+ * Copyright: Crownstone (https://crownstone.rocks)
+ * Date: Apr 16, 2021
+ * License: LGPLv3+, Apache License 2.0, and/or MIT (triple-licensed)
+ */
+
+#include <ble/cs_CrownstoneCentral.h>
+#include <ble/cs_BleCentral.h>
+#include <ble/cs_UUID.h>
+#include <cfg/cs_UuidConfig.h>
+#include <encryption/cs_ConnectionEncryption.h>
+#include <encryption/cs_KeysAndAccess.h>
+
+#include <structs/cs_ControlPacketAccessor.h>
+#include <structs/cs_ResultPacketAccessor.h>
+#include <structs/buffer/cs_CharacteristicBuffer.h>
+#include <structs/buffer/cs_CharacteristicReadBuffer.h>
+#include <structs/buffer/cs_CharacteristicWriteBuffer.h>
+#include <structs/buffer/cs_EncryptionBuffer.h>
+
+#define LOGCsCentralInfo LOGi
+#define LOGCsCentralDebug LOGv
+#define LogLevelCsCentralDebug SERIAL_VERBOSE
+
+cs_ret_code_t CrownstoneCentral::init() {
+	// Make use of the fact ERR_SUCCESS = 0, to avoid many if statements.
+	cs_ret_code_t retCode = ERR_SUCCESS;
+	retCode |= _uuids[ServiceIndex::SERVICE_INDEX_CROWNSTONE].fromFullUuid(CROWNSTONE_UUID);
+	retCode |= _uuids[ServiceIndex::SERVICE_INDEX_SETUP].fromFullUuid(SETUP_UUID);
+	retCode |= _uuids[ServiceIndex::SERVICE_INDEX_DEVICE_INFO].fromShortUuid(BLE_UUID_DEVICE_INFORMATION_SERVICE);
+	retCode |= _uuids[ServiceIndex::SERVICE_INDEX_DFU].fromShortUuid(0xFE59); // DFU service
+	if (retCode != ERR_SUCCESS) {
+		return retCode;
+	}
+	reset();
+	listen();
+	return ERR_SUCCESS;
+}
+
+void CrownstoneCentral::reset() {
+	_sessionKeyHandle =  BLE_GATT_HANDLE_INVALID;
+	_sessionDataHandle = BLE_GATT_HANDLE_INVALID;
+	_controlHandle =     BLE_GATT_HANDLE_INVALID;
+	_resultHandle =      BLE_GATT_HANDLE_INVALID;
+	_resultCccdHandle =  BLE_GATT_HANDLE_INVALID;
+	_opMode =            OperationMode::OPERATION_MODE_UNINITIALIZED;
+	resetNotifactionMergerState();
+}
+
+void CrownstoneCentral::resetNotifactionMergerState() {
+	_notificationNextIndex = 0;
+	_notificationMergedDataSize = 0;
+}
+
+cs_ret_code_t CrownstoneCentral::connect(stone_id_t stoneId, uint16_t timeoutMs) {
+	return ERR_NOT_IMPLEMENTED;
+}
+
+cs_ret_code_t CrownstoneCentral::connect(const device_address_t& address, uint16_t timeoutMs) {
+	if (isBusy()) {
+		LOGCsCentralInfo("Busy");
+		return ERR_BUSY;
+	}
+
+	cs_ret_code_t retCode = BleCentral::getInstance().connect(address, timeoutMs);
+	if (retCode == ERR_WAIT_FOR_SUCCESS) {
+		_currentOperation = Operation::CONNECT;
+		setStep(ConnectSteps::CONNECT);
+	}
+	return retCode;
+}
+
+cs_ret_code_t CrownstoneCentral::disconnect() {
+	if (isBusy()) {
```

Comment:
do we have 'force disconnect' kind of option?

Reply:
Hmm, good question. We might indeed be able to disconnect at any moment, as that might happen at any moment anyway.